panel.py

```diff
@@ -3,101 +3,106 @@
-  # notice how the action here is to "defenestrates the aristocracy", not merely "push the button".
-  # Another button might "launch the missiles". Two dials might respectively be labeled
-  # "Froomulator" and "Hypervisor", with actions "Set Froomulator to 1" vs. "Set Hypervisor to 1".
-  #
-  # See the third control below for a shorthand way to define `actions` for dial-like controls.
-  'actions': {
-    '0': '',
-    '1': 'Defenestrate the aristocracy!'
-  },
-
-    # HACK(jeff): `'type': 'button'` is only present to be able to play the game via the CLI, see
-    # where this is read at the bottom of this script.
-  'type': 'button',
-}, {
-  'id': 'octo',
-
-  'state': 'nothing',
-
-  'actions': {
-    'nothing': '',
-    'nipple': 'Octo bite raven girl nipple!',
-    'mouth': 'Octo kiss raven girl mouth!'
-  },
-
-  'type': 'button'
-}, {
-  'id': 'Froomulator',
-
-  'state': '0',
-
-  # This is a shorthand form of `states`, where the first value (the array) contains the values for
-  # `state`, and the second value (the string) is a "template action": the actual actions for each
-  # state will be formed by replacing "%s" with the state.
-  'actions': [
-    ['0', '1', '2'],
-    'Set Froomulator to %s!'
+# connect/disconnect, add and remove from this list, then call `announce`
+# again.
+controls = [
+    {
+      # An internal identifier for the control. Must be globally unique. Pick a
+      # name that's meaningful for you, this won't be seen by the player.
+      'id': 'defenestrator',
+
+      # Type -- must be one of 'button', 'switch', or 'selector'
+      'type': 'button',
+
+      # Value -- the current value of the control.
+      # The possible values depend on the type. For a button or a switch, `value` must
+      # be a boolean (one of the set of [True, False, 0, 1]).
+      # For a selector, value must be one of the values given as possible in a
+      # `values` key, which can be anything, e.g. `['0', '1', '2', '3']` or
+      # `['red', 'green', 'yellow']`.
+      'value': False,
+
+      # How will the switch be identified in instructions to players?
+      # This is probably a label on that section of the control panel
+      'description': 'the Aristocracy',
+
+      # What's the 'verb' of the control? This is probably a label on the control
+      # itself
+      'action': 'Defenestrate',
+    }, {
+      'id': 'girl-nipple',
+      'type': 'button',
+      'value': False,
+      'description': 'the raven girl\'s nipple',
+      'action': 'Octo bite',
+    }, {
+      'id': 'girl-mouth',
+      'type': 'button',
+      'value': False,
+      'description': 'the raven girl\'s mouth',
+      'action': 'Octo kiss',
+    }, {
+      'id': 'froom',
+      'type': 'selector',
+      'value': '0',
+      'description': 'Froomulator',
+      'action': 'Froomulate',
+      'values': [
+        '0',
+        '1',
+        '2',
+        ],
+    }
   ]
-}]
 
+class KeyboardReader(threading.Thread):
```

Comment:
I'm not opposed to these changes but curious, what does refactoring this do for us? :) The keyboard code is just for dev purposes, will be replaced for the real panels

Reply:
yeah there's no real reason for that